README.md
# Complex System Simulation - Financial Phenomena in Genoa market model



Group 14

Authors: Joos Akkerman, Oscar de Keijzer, Susy Maijer & Loek van Steijn



## Introduction
In this project we adopt the Genoa Market model by Raberto et al. [[1]](#1) to determine the influence of cluster/group formation amongst individual traders (i.e. Hedge Funds or groups of private investors) on commonly found emergent phenomena in financial markets. The model consists of a set of agents, which perform financial transactions among each other. As agents determine their limit price and desired quantity to be sold or bought, an equilibrium price is found that clears the market. During this process traders are linked together to form a cluster with probability $P_a$, these clusters can merge with other clusters if two agents from two different clusters are linked together. Every model iteration a cluster is randomly activated with probability $P_c$, when a cluster is activated each trader therein will either cumulatively want to sell or buy the asset after which the cluster is deleted. Over time, this generates a price series of which we analyse the commonly found emergent phenomena and compare it to the real-world financial time series of the S&P500 index and a Gaussian distributed return time series.

We primarily focus on the "cubic law of stock returns" [[2]](#2), which dictates that the fat tail of stock returns can be approximated as a power law distribution with an exponent of $\sim 3$. Secondly, we focus on the volatility clustering [[3]](#3), which states that large returns tend be followed by large returns and small returns tend to be followed by small returns. Lastly, we focus on the wealth distribution amongst individual traders, this is supposed to be Pareto distributed and can be summarized that roughly 80% of the total wealth is held by 20% of the total population.


## Navigation
Overview of all model elements and how to navigate them.

## Main
In this file the main model is executed and the main model parameters can be adjusted here too. The parameters include:

* `N_time`: integer, number of model iterations
* `cluster`: boolean, True if traders are allowed to form clusters, else Traders act independently
* `C`: integer, amount of cash each trader agent is initialized with
* `A`: integer, number of stocks each trader agent is initialized with
* `p`: integer, starting price of stock
* `garch`: boolean, True if volatility of stock is forecasted with Garch model
* `garch_n`: integer, Used in data storage
* `garch_param`: list of two intgers, Garch(p,q) degrees, standard [1,1]
* `Pa_list`: list of floats, contains different probabilities of two traders forming a cluster
* `Pc_list`: list of floats, contains different probabilities of a cluster activating where all traders therein cumulatively buy or sell stock
* `N_agents`: list of integers, contains different number of agents in model
* `N_concurrent`: integer, how many times each simulation is performed, increases statistical relevance
* `N_cores`: integer, Number of CPU cores maximally used for data collection


### Classes
Overview of all model classes.

#### Trader
The Trader class contains the data and methods for the individual trader objects. The following data is stored:

* `id`: id of the trader
* `market`: object, refers to the market object the trader is in.
* `C`: list of floats, time series of the cash owned by the trader
* `A`: list of floats, time series of assets owned by the trader
* `a_b`: float, quantity of assets the trader wants to buy at time `t`
* `b_i`: float, limit price trader wants to buy at (so highest possible price, will buy for a lower price in most cases).
* `a_s`: float, quantity of assets the trader wants to sell at time `t`
* `s_i`: float, limit price trader wants to sell at (so lowest possible price, will sell for a higher price in most cases).
* `P_i`: float, buy order probability. This determines if the trader will set a buy or sell order.
* `in_cluster`: object, refers to cluster object, if the trader is in a cluster, otherwise `None`.

The Trader object contains the following methods:

* `trade_decision()`: decides if trader will place buy or sell order, based on probability `P_i`. Then refers to `buy()` or `sell()`.
* `buy()`: determines `a_b` and `b_i`, places buy order.
* `sell()`: determines `a_s` and `s_i`, places sell order.
* `no_trade()`: if the market price does not meet the limit price, no trade is made. This function then extends the series for `A` and `C` with their respective values at time `t`.

#### Market
The Market object contains the methods that allow the market mechanism to function. It also tracks the macro variables of the market. The following data is stored:

* `p`: list of floats, time series of the market price.
* `cluster`: boolean, determines if clustering is activated.
* `traders`: list of objects, containing all trader objects
* `buyers`: list of objects, containing all buyers at time `t`, emptied after time step is over.
* `sellers`: list of objects, containing all sellers at time `t`, emptied after time step is over.
* `clusters`: list of objects, containing all clusters at time `t`. Updated if clusters are added or deleted.
* `Pc`: float, probability of clusters activating
* `Pa`: float, probability of clusters forming
* `hist_vol`: float, historical volatility. updated based on generated time series.
* `sigma`: list of floats, time series of true volatility, as the historical volatility is multiplied by a constant $3.5$, in accordance with [[1]](#1). Alternatively, this can be done using an implementation of a GARCH model (which was not used in the final results).
* `avg_degree`: list of floats, average degree of clusters.
* `garch`: boolean, governs whether a GARCH model is used to determine `sigma`.
* `garch_param`: list of integers, GARCH(p,q) degrees ([p,q]).
* `garch_model`: function, fitted GARCH model for time step `t`.

The market objects contains the following methods that determine the **pricing process**:

* `get_equilibrium_p()`: uses limit prices and quantities set by traders to determine the equilibrium price (for which supplied quantity equals demanded quantity). Utilizes `find_intersection()` to find intersection between supply and demand curves.
* `find_intersection()`: Finds set of buyers and sellers for which the market is cleared to the largest extend, and for which the determined equilibrium price satisfies the set limit price.
* `perform_transactions()`: receives true buyers and sellers, performs transactions by updating `A` and `C` for true buyers and sellers.

The market objects contains the following methods that determine the **volatility estimates**:

* `update_sigma()`: updates `sigma` series, using the updated `hist_vol` and either the method by [[1]](#1) or the fitted GARCH model.
* `update_hist_vol()`: updates `hist_vol` using the log-returns of the generated price series.
* `fit_GARCH()`: fits GARCH model to previous volatility and return series, ranging 20-100 time steps back, depending on availability of data. Determines volatility for time step `t`.

The market objects contains the following methods that determine the **clustering Process**:
* `form_clusters()`: decides which clusters are to be formed. Randomly chooses two traders each time step to be clustered. After which, it refers to the following functions based on the cluster state of the traders:
    * `init_cluster()`: if both traders are not in a cluster, form new cluster object. Update `clusters`.
    * `merge_cluster()`: if both traders are in a cluster, merge both clusters to a new cluster object. Update `clusters`.
    * If one of the traders is in a cluster and the other is not, add second trader to cluster (through method in `Cluster` object of first trader).

* `activate_cluster()`: based on `Pc`, activates one of the randomly chosen clusters.

Finally, the Market object contains the method `reset_lists()`, which clears the `buyers` and `sellers` lists.

#### Cluster
The Cluster class contains the data and methods for the cluster objects. The following data is stored:

* `members`: List of objects, refers to Trader objects that are part of the cluster.
* `market`: Object, refers to the Market object the cluster is part of.

The cluster object contains the following methods:
* `add_to_cluster()`: Adds a Trader object to the cluster.
* `activate()`: Activates cluster by setting all probabilities `P_i` to either 1 or 0 for all traders in the cluster.
* `self_destruct()`: After being activated, removes Cluster object from traders and Market object.


### Data
The file `data.py` contains some general utility functions

* `avg_degree()`: Calculates the average degree of the randomly formed clusters of traders at each model iteration
* `calc_norm_return()`: Calculates either the regular or absolute standardized log returns of financial time series
* `create_cdf()`: Creates cumulative distribution of any data by utilizing histograms
* `fit_func()`: Linear fit function of shape $y = a\cdot x + b$
* `curve_fit_log()`: Fits linear curve to function in a loglog basis
* `sample_gauss()`: Creates Gaussian distributed prices, based on a random variable with mean 100 and variance 0.1
* `SP500_pl()`: Creates cumulative distribution of SP500 returns over the past 30 years of closing data
* `vol_cluster()`: Quantify volatility clustering based on method by Tseng and Li [[4]](#4), method involves counting the number of days
    in a rolling window that exceed a certain return threshold.

### Management
The file `management.py` contains files used to store and retreive data

* `getResultsDirectoryPath()`: Returns the directory of the results of a specific run configuration.
* `makeDirectories()`: Make directories for each configuration parameter (done in advance due to concurrency problems which otherwise arise)
* `saveSingleMarket()`: Write the final Market Object to disk.
* `loadSingleMarket()`: Load a Market Object from disk.
* `loadMultipleMarkets()`: Load all N Market Objects from a single parameter configuration.


### Visualization

The file `visualization.py` contains the functions used to plot the results. These are the following:

* `vis_market_cross()`: visualizes the market equilibrium cross.![](https://i.imgur.com/tVkXxlr.png)

* `vis_price_series()`: Visualizes the return distribution of the price series, fits a power law to the (tail of the) distribution, shows the fitted curve with the best fit.
* `vis_wealth_over_time()` Visualizes the individual wealth time series of all agents, and the histogram of wealth at the final time step.
* `cluster_vis()`: visualizes the average degree of the formed clusters over time.![](https://i.imgur.com/OEmkR3F.png)

* `vis_vol_cluster()`: Visualizes volatility clustering graphs based on the method by Tseng and Li [[4]](#4) for multiple Pa, Pc and number of agents.![](https://i.imgur.com/gbE5JSB.png)


* `plot_wealth_dist()`: Visualizes the wealth distribution of single iteration.
* `plot_lorenz_curve()`: Visualizes the Lorenz curve and wealth distribution, averaged over multiple iterations for increasing time steps. This is to show the development of wealth distribution over time.![](https://i.imgur.com/3FyjvQy.png)


* ` plot_lorenz_curve_Nagents()`: Visualizes the Lorenz curve and wealth distribution, averaged over multiple iterations for increasing amount of agents. This is to show the development of wealth distribution as the number of agents increases.![](https://i.imgur.com/VcqQerQ.png)
* `vis_volatility_series()`: Visualizes time series of asset volatility.
* `visualiseSingleMarketResults()`: Calls plot functions to show the results of a single run.
* `visualiseMultipleMarketResults()`: Calls plot functions to show the results of multiple averaged runs.


## Operation
Before running any code it is important to update your python modules by installing requirements.txt found in the main folder. Ensure that you have Python and Pip installed. You can download Python [here](https://www.python.org/downloads/). You can install the neccessary Python modules by running:

```command
pip install -r requirements.txt
```

Once the correct packages are installed, any parameters can be changed in the `main.py` file as described above. The model can be executed to either collect data by running

```command
python main.py collect
```
This will save all objects to pickle. These files can be visualized by running

```command
python main.py visualize
```
This will visualise the collected data as described above.

## References

<a id="1">[1]</a> Raberto, M., Cincotti, S., Focardi, S. M., & Marchesi, M. (2001). Agent-based simulation of a financial market. _Physica A: Statistical Mechanics and its Applications_, _299_(1-2), 319-327.
<a id="2">[2]</a> Gopikrishnan, P., Plerou, V., Amaral, L.A.N., Meyer, M., Stanley, H.E., (1999) Scaling of the distribution of fluctuations of financial market indices. The American Physical Society Volume 60, Number 5
<a id="3">[3]</a> Mandelbrot, B., 1963. The variation of certain speculative prices. Journal of Business 36 (4), 394–419.
<a id="4">[4]</a> Tseng, J. J., & Li, S. P. (2012). Quantifying volatility clustering in financial time series. International Review of Financial Analysis, 23, 11-19.


Presentation/changelog.txt
Version: 4.3.0
  - Updated Bootstrap to version 5.0.1
  - Updated all outdated third party vendor libraries to their latest versions
  - Fixed navigation links focus color

Version: 4.2.0
  - Updated Bootstrap to version 5.0.0 Final
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.1.0
  - Updated Bootstrap to version 5.0.0-beta3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to V3.1

Version: 4.0.1
  - Updated Bootstrap to version 5.0.0-beta2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.0.0
  - The template does not require jQuery anymore
  - Removed jQuery and all the jQuery plugins
  - The assets/js/main.js was rewritten completely with vanilla Javascript. No more jQuery code
  - Restructured the dev version for better development experience
  - Updated the PHP Email Form to V3.0 - No jQuery dependency. Added attachment support

Version: 3.0.0
  - Initial release with Bootstrap v5.0 Beta 1

Version: 2.2.0
  - Updated Bootstrap to version 4.5.3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to v2.3
  - Other small fixes and improvements

Version: 2.1.0
  - Updated Bootstrap to version 4.5.0
  - Updated the PHP Email Form library to version 2.0 with reCaptcha support
  - Aded inner-page.html tempalte
  - Added smooth scroll on page load with hash links in the url
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
  - Initial Release


data.py
import yfinance as yf
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit
import pandas as pd

def avg_degree(MarketObj, cluster):
    """
    Determines the average degree of the random graph formed by the trader agents,
    saves average degree of current time step in market object
    @param MarketObj         Market object containing list of formed clusters
    @param cluster           Boolean to determine if model contains cluster formation
    """
    if cluster:
        degree_list = list(map(lambda x: len(x.members), MarketObj.clusters))
        MarketObj.avg_degree.append(np.mean(degree_list))

def calc_norm_return(df, abs):
    """
    Calculates standardized log returns of stock or index
    @param df        DataFrame containing closing prices of stock or index
    @param abs       Boolean to determine if absolute returns are required
    return:          DataFrame containing the standardized log returns
    """
    df = (np.log(df) - np.log(df.shift(1)))
    df = df.dropna()
    df = (df - df.mean())/df.std()
    if abs == True:
        df = df.abs()
    return df

def create_cdf(df):
    """
    Creates cumulative distribution of any data by utilizing histogram
    @param df       DataFrame containing desired data
    return:         Cumulative ditribution and histogram bin count
    """
    count, bins_count = np.histogram(df, bins=50)

    pdf = count / sum(count)
    pdf = np.flip(pdf)

    cdf = np.cumsum(pdf)
    cdf = np.flip(cdf)

    return cdf, bins_count

def fit_func(x, a, b):
    """
    Linear fit function used to fit onto a Power law distribution in form y = a + b*x
    @param a        float, constant
    @param b        float, constant
    @param x        float, input variable
    return:         a + b*x
    """
    return a + x * b

def curve_fit_log(xdata, ydata):
    """
    Fits linear curve to function in a loglog basis
    @param xdata        list of floats to serve as x data
    @param ydata        list of floats to serve as y data
    return:             list of fit params, list of parameter covariances, list of fitted y data
    """
    x_log = np.log10(xdata)
    y_log = np.log10(ydata)

    popt_log, pcov_log = curve_fit(fit_func, x_log, y_log)

    yfit_log = np.power(10, fit_func(x_log, *popt_log))

    return (popt_log, pcov_log, yfit_log)

def sample_gauss(N_time):
    """
    Creates Gaussian distributed prices, based on a random variable with mean 100 and variance 0.1,
    determines cumulative distribution of its standardized log returns.
    @param N_time       Integer length of time series
    return:             bins count of histogram, cumulative distribution
    """
    # creates Gaussian price series
    sample_list = list(map(lambda x: np.random.normal(100, 0.1), range(N_time)))
    series = pd.DataFrame(sample_list)
    # Determines standardized absolute returns
    series_norm = calc_norm_return(series, True)
    # create CDF
    cdf, bins_count = create_cdf(series_norm)

    return bins_count, cdf

def SP500_pl():
    """
    Creates cumulative distribution of SP500 returns over the past 30 years of closing data
    return:         histogram bin bount, cumulative distribution of SP500 returns
    """
    # retrieves data
    ticker = '^GSPC'
    df = yf.Ticker(ticker).history(period='30y', interval='1d')['Close']
    # standardized returns
    df = calc_norm_return(df, True)
    # cumulative distribution of returns and bin count of histogram
    cdf, bins_count = create_cdf(df)

    return bins_count, cdf

def vol_cluster(ret, highp, window, N_time, gauss):
    """
    Quantify volatility clustering based on method by Tseng and Li, method involves counting the number of days
    in a rolling window that exceed a certain return threshold.
    @param ret          list of float returns from the model
    @param highp        float between 0 and 1, percentage of highest returns to set as threshold
    @param window       integer, rolling window size
    @param N_time       integer, number of model iterations
    @param  gauss       boolean, if True method is applied to Guassian distributed returns and SP500 returns, else model returns
    return:             list of occurences where the return exceeded the threshold in the rolling window, equal to length N_time
    """
    if gauss:
        # gets Gaussian distribution price_returns_multiple_times
        sample_list = list(map(lambda x: np.random.normal(100, 0.1), range(N_time)))
        df = pd.DataFrame(sample_list)
        df = calc_norm_return(df, True)

        # defines threshold
        sample_sort = sorted(df.values, key=float)
        threshold_gaus = sample_sort[int((1-highp) * len(sample_sort))]

        # moves rolling window and counts occurences where threshold is exceeded
        cluster_gaus = [0 for _ in range(len(sample_list) - window)]
        for i in range(len(sample_list) - window):
            for j in range(window):
                if df.values[i+j] >= threshold_gaus:
                    cluster_gaus[i] += 1

        # repeats process for SP500 returns
        ticker = '^GSPC'
        df = yf.Ticker(ticker).history(period='30y', interval='1d')['Close']
        df = calc_norm_return(df, True)
        sp500_sort = sorted(df.values, key=float)
        threshold_sp500 = sp500_sort[int((1-highp) * len(sp500_sort))]

        cluster_sp500 = [0 for _ in range(len(df.values)- window)]
        for i in range(len(df.values) - window):
            for j in range(window):
                if df.values[i+j] >= threshold_sp500:
                    cluster_sp500[i] += 1

        return cluster_gaus, cluster_sp500

    else:
        # determines threshold for model returns
        ret_sorted = sorted(ret, key=float)
        threshold = ret_sorted[int((1-highp) * len(ret_sorted))]

        # moves rolling window and counts occurences where threshold is exceeded
        cluster_series = [0 for _ in range(len(ret)-window)]
        for i in range(len(ret)-window):
            for j in range(window):
                if ret[i+j] >= threshold:
                    cluster_series[i] += 1

        return cluster_series


trader.py
import numpy as np
import math

class Trader():
    def __init__(self, id, market, A, C):
        """
        Initialise trader with a specific amount of cash and assets.

        @param id       Unique trader id
        @param market   Reference to the market
        @param A        Initial amount of stocks
        @param C        Initial cash amount
        """
        self.id = id
        self.market = market # add market object for easy access to parameters
        self.C = [C] # initialize list for cash time series
        self.A = [A] # initialize list for stock time series
        self.a_b = 0 # quantity of stocks to buy
        self.b_i = 0 # limit buy price of trader's buy order
        self.a_s = 0 # quantity of stock offered for sale
        self.s_i = 0 # limit sell price of trader's sell order
        self.P_i = 0.5 # buy order probability
        self.in_cluster = None


    def trade_decision(self): 
        """
        Determine whether the trader wants to buy or sell at this time step.
        """
        if np.random.random() > self.P_i:
            self.market.buyers += [self]
            self.buy()
        else:
            self.market.sellers += [self]
            self.sell()


    def buy(self):
        """
        Trader performs a buy action.
        """

        # Generate buy limit price
        b_i = max(0, self.market.p[-1]*np.random.normal(self.market.mu, self.market.sigma[-1]))

        # Generate buy limit price
        if b_i == 0:
            self.a_b = 0
            self.b_i = 0
        else:
            # Determine amount of stocks to buy (spend random fraction of total cash)
            a_b = max(0, math.trunc(np.random.random() * self.C[-1] / b_i))

            # Set the buy limit price
            self.b_i = b_i
            self.a_b = a_b


    def sell(self):
        """
        Trader performs a sell action.
        """
        # Amount of stocks to sell is a random fraction of total stocks
        a_s = max(0, math.trunc(np.random.random() * self.A[-1]))

        # Generate sell limit price
        s_i = max(0, self.market.p[-1]/np.random.normal(self.market.mu, self.market.sigma[-1]))

        # Set the sell limit price and amount of stocks to cell
        self.s_i = s_i
        self.a_s = a_s


    def no_trade(self):
        """
        Called when no trade is made, in order to keep the asset and cash series consistent
        """
        self.A += [self.A[-1]]
        self.C += [self.C[-1]]


cluster.py
import numpy as np

class Cluster():
    def __init__(self, members, market):
        """
        Initialise cluster with a pair of traders.

        @param members          List of traders.         
        @param market           Market object.
        """
        self.members = members
        self.market = market


    def add_to_cluster(self, new_member):
        """
        Add new trader to a cluster.

        @PARAM member           Trader.
        """
        self.members += [new_member]
        new_member.in_cluster = self


    def activate(self):
        """
        Sets all probabilities to 0 or 1
        """
        cluster_P = np.random.choice([0,1])
        for member in self.members:
            member.Pi = cluster_P


    def self_destruct(self):
        """
        Resets probabilities of members and self destructs
        """
        for member in self.members:
            member.Pi = 0.5
            member.in_cluster = None

        self.market.clusters.remove(self)


management.py
import pickle
import os
from collections import Mapping, Container
import sys


def getResultsDirectoryPath(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster):
    """"
    Returns the directory of the results of a specific run configuration.

    @param N_agents             Amount of agents
    @param N_time               Amount of timesteps
    @param C                    Initial cash amount
    @param A                    Initial asset amount
    @param p                    Initial asset price
    @param garch                Whether to determine the volatility using GARCH.
    @param garch_n          
    @param garch_param
    @param Pa                   Probability of activating a cluster
    @param Pc                   Probability of forming a pair between two agents, i.e clustering probability
    @param cluster              Whether clustering is present or not
    """
    return f'results/agents{N_agents}_time{N_time}_C{C}_A{A}_p{p}_garch{garch}_garchN{garch_n}_garchparam{garch_param[0]}{garch_param[1]}_Pa{Pa}_Pc{Pc}_cluster{cluster}'


def makeDirectories(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list, Pc_list, cluster):
    """
     Make directories for each configuration parameter (done in advance due to concurrency problems which otherwise arise)

     @param Pa_list             List of cluster activation probabilities
     @param Pc_list             List of clustering probabilities.
     For further parameters, see function getResultsDirectoryPath()
    """

    # Loop over all possible parameter configurations
    for Pa in Pa_list:
        for Pc in Pc_list:
            # Get the directory of this parameter configuration
            results_dir = getResultsDirectoryPath(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster)

            # Make the directory if it doesn't exist yet
            if not(os.path.isdir(results_dir)):
                os.mkdir(results_dir)


def saveSingleMarket(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster, i, MarketObj):
    """
    Write the final MarketObject to disk.

    @param i                    The run number (i.e, 0 to 49)
    @param MarketObj            The MarketObj at the end of the run
    For further parameters, see function getResultsDirectoruPath()
    """
    
    #Make directory to save results from this simulation
    results_dir = getResultsDirectoryPath(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster)

    # Save results
    with open(f'{results_dir}/{i}.pickle', 'wb') as f:
        pickle.dump(MarketObj, f)


def loadSingleMarket(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster, i):
    """
    Load a MarketObject from disk.

    @param i                    The run number (i.e, 0 to 49)
    For further parameters, see function getResultsDirectoruPath()
    """

    # Get directory path
    results_dir = getResultsDirectoryPath(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster)

    # Load results
    with open(f'{results_dir}/{i}.pickle', 'rb') as f:
        MarketObj = pickle.load(f)

    return MarketObj


def loadMultipleMarkets(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster, N):
    """
    Load all N MarketObjects from a single parameter configuration.

    @param N                    The amount of simulation results.
    For further parameters, see function getResultsDirectoruPath()
    """

    # Load N runs
    objects = []
    for i in range(0, N):
        MarketObj = loadSingleMarket(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster, i)
        objects.append((MarketObj, [cluster, N_agents, Pa, Pc]))

    return objects


main.py
import numpy as np
from tqdm import tqdm
import concurrent.futures
import sys

# Our own defined modules
import gc
import market
import trader
import visualization as vis
import management
from data import avg_degree


def initialise(N_agents, p, A, C, cluster, garch, garch_param, Pa, Pc):
    """
    Initialise simulation run with market and traders.

    @param N_agents             Amount of agents
    @param p                    Initial asset price
    @param C                    Initial cash amount
    @param cluster              Whether clustering is present or not
    @param A                    Initial asset amount
    @param garch                Whether to determine the volatility using GARCH.
    @param garch_n
    @param garch_param
    @param Pa                   Probability of activating a cluster
    @param Pc                   Probability of forming a pair between two agents, i.e clustering probability
    """
    # Initialise market object
    MarketObj = market.Market(p, cluster, garch, garch_param, Pa, Pc)

    # Initialise agents in the market
    for i in range(N_agents):
        TraderObj = trader.Trader(i, MarketObj, A, C)
        MarketObj.traders += [TraderObj]

    return MarketObj


def run_simulation(N_time, MarketObj, cluster):
    """
    Run the Genoa Market Model once.

    @param N_time               Amount of timesteps
    @param MarketObj            Initial market
    @param cluster              Whether clustering is present or not
    """
    for t in tqdm(range(N_time)):

        # Perform clustering steps
        if cluster:
            MarketObj.form_clusters()
            activated_cluster = MarketObj.activate_cluster()

        # All traders place buy or sell order
        for TraderObj in MarketObj.traders:
            TraderObj.trade_decision()

        # Get average clustering degree
        avg_degree(MarketObj, cluster)

        # Determine clearing price and get viable sellers and buyers
        transaction_q, true_sellers, true_buyers = MarketObj.get_equilibrium_p()

        # COMMENT OUT TO VISUALISE DETERMINATION OF CLEARING PRICE
        vis.vis_market_cross(MarketObj, transaction_q)

        # Perform transactions and update the historic volatility
        MarketObj.perform_transactions(transaction_q, true_sellers, true_buyers)
        MarketObj.update_hist_vol()

        # Reset the activated cluster (if there was one)
        if cluster and activated_cluster != None:
            activated_cluster.self_destruct()

    return MarketObj

def job(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list, Pc_list, cluster, i):
    """
    Parallel programming function. It performs a single simulation run for each parameter configuration.

    @param Pa_list              List of cluster activation probabilities
    @param Pc_list              List of clustering probabilities.
    For further parameters, see function getResultsDirectoryPath()
    """
    # Force new seed, otherwise 6 processes (amount of physical cores) will use same seed
    np.random.seed()

    # Loop over every parameter configuration
    for Pa in Pa_list:
        for Pc in Pc_list:
            for N_agent in N_agents:

                # Run single simulation for this parameter config
                MarketObj = initialise(N_agent, p, A, C, cluster, garch, garch_param, Pa, Pc)
                MarketObj = run_simulation(N_time, MarketObj, cluster)

                # Save the results form this run
                management.saveSingleMarket(N_agent, N_time, C, A, p, garch, garch_n, garch_param, Pa, Pc, cluster, i, MarketObj)

                # Delete market object
                del MarketObj
                gc.collect()

    return " Done"

if __name__ == '__main__':

    if len(sys.argv) != 2:
        print("Execute 'python main.py collect' to collect data or 'python main.py visualize' for visualization of stored data.")
        sys.exit()

    # N_time: model iterations, cluster: use agents clustering, C: starting capital of agents, A: starting assets of agents, p: starting asset price.
    N_time = 10000
    cluster = True
    C = 30000
    A = 300
    p = 100

    # Garch parameters
    garch = False
    garch_n = 4
    garch_param = [1,1]

    # Experiment ranges
    Pa_list = [0.0001, 0.0002, 0.0005]
    Pc_list = [0.1, 0.2, 0.3]
    N_agents = [100, 200, 400]

    # Amount of runs per configuration
    N_concurrent = 2 ## NOTE: this was 50 before but this takes a long time
    # number of CPU cores maximally used to collect data
    N_cores = 6

    # Make directories for each parameter configuration (if they don't exist yet). NB don't comment this out
    management.makeDirectories(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list, Pc_list, cluster)

    # Do experiments for all Pa and Pc parameter combinations
    if sys.argv[1] == "collect":
        with concurrent.futures.ProcessPoolExecutor(max_workers=N_cores) as executor:
           values = [executor.submit(job, N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list, Pc_list, cluster, i,) for i in range(0, N_concurrent)]

    if sys.argv[1] == "visualize":
        # Visualisation all model runs of all parameter configurations
        vis.visualiseMultipleMarketResults(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list, Pc_list, cluster, N_concurrent)
        # Visualisation single model run (first run)
        vis.visualiseSingleMarketResults(N_agents, N_time, C, A, p, garch, garch_n, garch_param, Pa_list[1], Pc_list[0], cluster, 0)

    else:
        print("Execute 'python main.py collect' to collect data or 'python main.py visualize' for visualization of stored data.")
        sys.exit()
